### PR TITLE
feat: create workspace from loading existing exported file

### PR DESCRIPTION
**Pull request by pengx17**
This pull request adds the feature to export and import workspaces as SQLite files, refactors the IPC communication between the main and renderer processes in the Electron app, and fixes some bugs.

### Data flow design

image

### How does adding workspace work behind the scene
image

### How does main registers handlers?
image

New workspace:
<video src="https://user-images.githubusercontent.com/584378/235758141-0c633fee-5cca-4f0c-94dd-b7e0d17ffc03.mp4"></video>

Add workspace (import from existing db file):
<video src="https://user-images.githubusercontent.com/584378/235758129-c8af7436-6ad0-4f8d-b764-93a67b16d007.mp4"></video>

New workspace general setting
a8305f57-684b-4cb8-9a0c-ebc5262e466c

### CODE REVIEW

File: apps/electron/layers/main/src/events/register.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { app,BrowserWindow } from 'electron';
```
Comment (Brooooooklyn) [inline apps/electron/layers/main/src/events/register.ts:1]: How does this line pass the lint & format?

File: apps/electron/layers/main/src/exposed.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { allEvents as events } from './events';
+import { allHandlers as handlers } from './handlers';
+
+// this will be used by preload script to expose all handlers and events to the renderer process
+export { events,handlers };
```
Comment (Brooooooklyn) [inline apps/electron/layers/main/src/exposed.ts:5]: `prettier`?

File: apps/electron/layers/main/src/events/register.ts (hunk 1 above)
Comment (pengx17) [reply apps/electron/layers/main/src/events/register.ts:1]: No idea 🥵

**Comment by pengx17** [comment]
@Himself65 should be ready for review

File: apps/electron/layers/main/src/events/register.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+function getActiveWindows() {
+  return BrowserWindow.getAllWindows().filter(win => !win.isDestroyed());
+}
+
+export function registerEvents() {
+  // register events
+  for (const [namespace, namespaceEvents] of Object.entries(allEvents)) {
+    for (const [key, eventRegister] of Object.entries(namespaceEvents)) {
+      const subscription = eventRegister((...args) => {
+        const chan = `${namespace}:${key}`;
```
Comment (himself65) [inline apps/electron/layers/main/src/events/register.ts:21]: what is `chan`

File: apps/electron/layers/main/src/handlers/db/ensure-db.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
+import { watch } from 'chokidar';
```
Comment (himself65) [inline apps/electron/layers/main/src/handlers/db/ensure-db.ts:1]: This project has not been maintained for a long time. I'm worried about this

https://github.com/paulmillr/chokidar

Why don't we use Node.js API?

File: apps/electron/layers/main/src/handlers/db/sqlite.ts [hunk 5]
```diff
@@ -33,75 +34,100 @@ interface BlobRow {
...
     }
 
     // use cached version?
-    const db = (this.sqliteDB = sqlite(this.path));
+    const db = (this.db = sqlite(this.path));
     db.exec(schemas.join(';'));
 
     if (!this.firstConnect) {
-      this.ydoc.on('update', this.addUpdateToSQLite);
+      this.ydoc.on('update', (update: Uint8Array, origin) => {
```
Comment (himself65) [inline apps/electron/layers/main/src/handlers/db/sqlite.ts:70]: what about `off`?

File: apps/electron/layers/main/src/events/register.ts (hunk 3 above)
Comment (pengx17) [reply apps/electron/layers/main/src/events/register.ts:21]: chan === channel

File: apps/electron/forge.config.js [hunk 6]
```diff
@@ -45,6 +45,7 @@ module.exports = {
           teamId: process.env.APPLE_TEAM_ID,
         }
       : undefined,
+    // do we need the following line?
```
Comment (m1911star) [inline apps/electron/forge.config.js:48]: yes for prod version

File: apps/electron/layers/main/src/handlers/db/ensure-db.ts (hunk 4 above)
Comment (pengx17) [reply apps/electron/layers/main/src/handlers/db/ensure-db.ts:1]: I tried to use Node.js api first, but it does not work pretty well.

On the README of chokidar, it listed some issues:

```
Why?
Node.js fs.watch:

Doesn't report filenames on MacOS.
Doesn't report events at all when using editors like Sublime on MacOS.
Often reports events twice.
Emits most changes as rename.
Does not provide an easy way to recursively watch file trees.
Does not support recursive watching on Linux.

Node.js fs.watchFile:

Almost as bad at event handling.
Also does not provide any recursive watching.
Results in high CPU utilization.
Chokidar resolves these problems.
```

I do encounter most of the issues when using either `fs.watch` or `fs.watchFile`. For a more reliable implementation, I used the popular chokidar instead.

BTW do you have any other suggestions over chokidar?
I saw in https://github.com/parcel-bundler/parcel/issues/863 it listed some alternatives like: `watchman`, `fb-watchman` and `nsfw`. Not sure which one is better.

File: apps/electron/layers/main/src/handlers/db/sqlite.ts (hunk 5 above)
Comment (pengx17) [reply apps/electron/layers/main/src/handlers/db/sqlite.ts:70]: I think the listener will be gc'ed after `ydoc` is destoryed, thus there is no need to unlisten it expclitly.

File: apps/electron/layers/main/src/handlers/db/ensure-db.ts (hunk 4 above)
Comment (himself65) [reply apps/electron/layers/main/src/handlers/db/ensure-db.ts:1]: Let's move to a better package in the next PR.

**Review by Himself65 — APPROVED** [review]

**Pull request merged**